client/src/socket.js

```diff
@@ -31,14 +32,164 @@ class ProtocolError extends Error {
+    this.handshake = new AsyncSubject()
+    this.handshakeMessage = handshakeMessage
+    this.handshakeSub = null
+
+    // This is used to emit status changes that others can hook into.
+    this.status = new BehaviorSubject(STATUS_UNCONNECTED)
+    // Keep track of subscribers so we's can decide when to
+    // unsubscribe.
+    this.activeSubscribers = 0
+    this.requestCounter = 0
+    this.keepalive = keepalive
+    // A map from request_ids to an object with metadata about the
+    // request. Eventually, this should allow re-sending requests when
+    // reconnecting.
+    this.activeRequests = new Map()
+  }
+
+  deactivateRequest(requestId) {
+    this.activeRequests.delete(requestId)
+    if (this.activeRequests.size === 0) {
+      // Unsubscribe handshake and socket
+      this.unsubscribe()
+    }
+  }
+
+  // This is a trimmed-down version of multiplex that only listens for
+  // the handshake requestId, and doesn't send `end_subscription`
+  // messages etc that aren't needed. It also starts the keepalive
+  // observable and cleans up after it when the handshake is cleaned
+  // up.
+  sendHandshake() {
+    const requestId = this.requestCounter++
+    const handshake = this.handshake
+    const status = this.status
+
+    const request = Object.assign(
+      { request_id: requestId }, this.handshakeMessage)
+
+    this.next(request)
+    this.activeRequests.set(requestId, { request })
+    this.handshakeSub = this.subscribe({
+      next: incomingMsg => {
+        // Check if we're getting a response to the handshake or not
+        if (incomingMsg.request_id === requestId) {
+          // This is a per-query error
+          if (incomingMsg.error) {
+            handshake.error(incomingMsg)
+            status.next(STATUS_ERROR)
+          } else {
+            handshake.next(incomingMsg)
+            handshake.complete()
+            status.next(STATUS_READY)
+          }
+        }
+      },
+      error: err => {
+        handshake.error(err)
+        status.next(STATUS_ERROR)
+      },
+      complete: () => handshake.complete(),
+    })
+    this.handshakeSub.add(() => {
+      this.activeRequests.delete(requestId)
+    })
+    // Start the keepalive. We don't do anything with the subscription
+    // since we don't care about responses. All that matters is that
+    // we can stop it later when needed. If an error to a keepalive
+    // comes back, that's fine since either the socket is down (which
+    // we'll find out about elsewhere) or the keepalive "failed" which
+    // is a server bug, but indicates the socket is still alive, so
+    // it's not really a failure.
+    const keepAliveSub = Observable
+            .timer(this.keepalive * 1000, this.keepalive * 1000)
+            .map(n => ({ type: 'keepalive', n }))
+            .subscribe()
+    // Make sure keepalive is killed when the handshake is cleaned up
+    this.handshakeSub.add(keepAliveSub)
+  }
+
+  multiplex(rawRequest) {
```

Comment:
They are intentionally wrapping the `tryCatch` in a closure because that's a fairly well-known workaround for ensuring the function can be optimized by the V8 optimizing compiler, meaning it has minimal performance impact at runtime (`bluebird` uses the same trick). I would imagine bytes saved is the same or better in terms of performance than reducing closures in this case.

The only real advantage is that it's easier to read and thus make changes. I think we probably can/should do the same thing in `sendHandshake()` as well.